src/Constraint/MultiConstraint.php

```diff
@@ -31,6 +31,8 @@ class MultiConstraint implements ConstraintInterface
     /** @var Bound|null */
     protected $upperBound;
 
+    private $cacheString;
```

Comment:
This cache of `__toString` should be extracted to a separate PR IMO, as it also benefits other optimization work.